test/EndToEndTests/Microsoft.OData.E2E.Portable/AsynchronousTests/AsynchronousQueryTests.cs

```diff
@@ -1197,6 +1197,36 @@ public void Linq_Where_GreaterThan()
             this.WaitForTestToComplete();
         }
 
+        /// <summary>
+        /// When DataServiceContext.MakeIdPredicateByKey=false, An expression with an Id as a predicate should create an Uri with a $filter in query options
+        /// </summary>
+        [Fact]
+        public void Linq_Where_Generates_Filter_By_Id_When_MakeIdPredicateByKey_Is_False()
+        {
+            var context = this.CreateWrappedContext<DefaultContainer>().Context;
+            context.MakeIdPredicateByKey = false;
+            var query = (from c in context.Customer
+                         where c.CustomerId == -10
+                         select new Customer { Name = c.Name, CustomerId = c.CustomerId }) as DataServiceQuery<Customer>;
+            var uri = query.ToString();
+            Assert.Contains("$filter=CustomerId eq -10", uri);
+        }
+
+        /// <summary>
+        /// When DataServiceContext.MakeIdPredicateByKey=true, An expression with an Id as a predicate should create a ByKey Uri
+        /// </summary>
+        [Fact]
+        public void Linq_Where_Generates_Filter_By_Id_When_MakeIdPredicateByKey_Is_True()
+        {
+            var context = this.CreateWrappedContext<DefaultContainer>().Context;
+            context.MakeIdPredicateByKey = true;

```

Comment:
@KenitoInc @habbes I think there's a need to rethink this, starting with the naming for this new property. Based on the current behaviour, a predicate with key field results into a ByKey query. Whatever new property we introduce, we need a value of `false` (default for boolean) to map to this current behaviour. This makes me think that the new property should be named in a way that it conveys the meaning "Id predicate should generate a filter query". In other words the reverse of the property that was added - something like `IdPredicateGeneratesFilterQueryOption` (in not such tacky form).

Reply:
I think your "not so tacky" form is actually more readable and unambiguous than the shorter "MakeIdPredicateByKey". I think @KenitoInc opted to for the value to be `true` when reflecting the current behaviour and then setting it to `true` by default. In my opinion either way works so long as the default value reflects the current behaviour.

Reply:
#Resolved